File: regions/system/library/typescript/scheduler-client/__tests__/scheduler-client.test.ts
import { describe, it, expect } from 'vitest';
import {
  InMemorySchedulerClient,
  SchedulerError,
  type JobRequest,
} from '../src/index';

describe('InMemorySchedulerClient', () => {
  it('should create a job', async () => {
    const client = new InMemorySchedulerClient();
    const req: JobRequest = {
      name: 'daily-report',
      schedule: { type: 'cron', expression: '0 2 * * *' },
      payload: { reportType: 'daily' },
      maxRetries: 3,
      timeoutSecs: 300,
    };
    const job = await client.createJob(req);
    expect(job.id).toBe('job-001');
    expect(job.name).toBe('daily-report');
    expect(job.status).toBe('pending');
  });

  it('should cancel a job', async () => {
    const client = new InMemorySchedulerClient();
    const job = await client.createJob({
      name: 'test',
      schedule: { type: 'cron', expression: '* * * * *' },
      payload: {},
    });
    await client.cancelJob(job.id);
    const got = await client.getJob(job.id);
    expect(got.status).toBe('cancelled');
  });

  it('should pause and resume a job', async () => {
    const client = new InMemorySchedulerClient();
    const job = await client.createJob({
      name: 'test',
      schedule: { type: 'cron', expression: '* * * * *' },
      payload: {},
    });
    await client.pauseJob(job.id);
    expect((await client.getJob(job.id)).status).toBe('paused');

    await client.resumeJob(job.id);
    expect((await client.getJob(job.id)).status).toBe('pending');
  });

  it('should throw on get non-existent job', async () => {
    const client = new InMemorySchedulerClient();
    await expect(client.getJob('nonexistent')).rejects.toThrow(SchedulerError);
  });

  it('should list jobs with filter', async () => {
    const client = new InMemorySchedulerClient();
    await client.createJob({
      name: 'job-a',
      schedule: { type: 'cron', expression: '* * * * *' },
      payload: {},
    });
    const jobB = await client.createJob({
      name: 'job-b',
      schedule: { type: 'one_shot', runAt: new Date() },
      payload: {},
    });
    await client.pauseJob(jobB.id);

    const paused = await client.listJobs({ status: 'paused' });
    expect(paused).toHaveLength(1);
    expect(paused[0].status).toBe('paused');
  });

  it('should filter by name prefix', async () => {
    const client = new InMemorySchedulerClient();
    await client.createJob({
      name: 'daily-report',
      schedule: { type: 'cron', expression: '0 2 * * *' },
      payload: {},
    });
    await client.createJob({
      name: 'weekly-report',
      schedule: { type: 'cron', expression: '0 2 * * 0' },
      payload: {},
    });

    const daily = await client.listJobs({ namePrefix: 'daily' });
    expect(daily).toHaveLength(1);
    expect(daily[0].name).toBe('daily-report');
  });

  it('should return empty executions', async () => {
    const client = new InMemorySchedulerClient();
    const execs = await client.getExecutions('job-001');
    expect(execs).toHaveLength(0);
  });

  it('should return all jobs', async () => {
    const client = new InMemorySchedulerClient();
    await client.createJob({
      name: 'a',
      schedule: { type: 'interval', intervalMs: 60000 },
      payload: {},
    });
    expect(client.getAll()).toHaveLength(1);
  });

  it('should throw on cancel non-existent job', async () => {
    const client = new InMemorySchedulerClient();
    await expect(client.cancelJob('none')).rejects.toThrow(SchedulerError);
  });

  it('should use default maxRetries and timeoutSecs', async () => {
    const client = new InMemorySchedulerClient();
    const job = await client.createJob({
      name: 'defaults',
      schedule: { type: 'cron', expression: '* * * * *' },
      payload: {},
    });
    expect(job.maxRetries).toBe(3);
    expect(job.timeoutSecs).toBe(60);
  });
});

File: regions/system/library/dart/scheduler_client/test/scheduler_client_test.dart
import 'package:k1s0_scheduler_client/scheduler_client.dart';
import 'package:test/test.dart';

void main() {
  group('InMemorySchedulerClient', () {
    test('ジョブが作成されること', () async {
      final client = InMemorySchedulerClient();
      final job = await client.createJob(JobRequest(
        name: 'daily-report',
        schedule: Schedule.cron('0 2 * * *'),
        payload: {'report_type': 'daily'},
        maxRetries: 3,
        timeoutSecs: 300,
      ));
      expect(job.id, equals('job-001'));
      expect(job.name, equals('daily-report'));
      expect(job.status, equals(JobStatus.pending));
    });

    test('ジョブがキャンセルされること', () async {
      final client = InMemorySchedulerClient();
      final job = await client.createJob(JobRequest(
        name: 'test',
        schedule: Schedule.cron('* * * * *'),
      ));
      await client.cancelJob(job.id);
      final got = await client.getJob(job.id);
      expect(got.status, equals(JobStatus.cancelled));
    });

    test('ジョブが一時停止および再開されること', () async {
      final client = InMemorySchedulerClient();
      final job = await client.createJob(JobRequest(
        name: 'test',
        schedule: Schedule.cron('* * * * *'),
      ));
      await client.pauseJob(job.id);
      expect((await client.getJob(job.id)).status, equals(JobStatus.paused));

      await client.resumeJob(job.id);
      expect((await client.getJob(job.id)).status, equals(JobStatus.pending));
    });

    test('存在しないジョブ取得時に例外がスローされること', () async {
      final client = InMemorySchedulerClient();
      expect(() => client.getJob('nonexistent'), throwsA(isA<SchedulerError>()));
    });

    test('ステータスフィルターでジョブ一覧が取得できること', () async {
      final client = InMemorySchedulerClient();
      await client.createJob(JobRequest(
        name: 'job-a',
        schedule: Schedule.cron('* * * * *'),
      ));
      final jobB = await client.createJob(JobRequest(
        name: 'job-b',
        schedule: Schedule.cron('* * * * *'),
      ));
      await client.pauseJob(jobB.id);

      final paused =
          await client.listJobs(const JobFilter(status: JobStatus.paused));
      expect(paused.length, equals(1));
      expect(paused[0].status, equals(JobStatus.paused));
    });

    test('名前プレフィックスフィルターでジョブ一覧が取得できること', () async {
      final client = InMemorySchedulerClient();
      await client.createJob(JobRequest(
        name: 'daily-report',
        schedule: Schedule.cron('0 2 * * *'),
      ));
      await client.createJob(JobRequest(
        name: 'weekly-report',
        schedule: Schedule.cron('0 2 * * 0'),
      ));

      final daily =
          await client.listJobs(const JobFilter(namePrefix: 'daily'));
      expect(daily.length, equals(1));
      expect(daily[0].name, equals('daily-report'));
    });

    test('実行履歴が空で返ること', () async {
      final client = InMemorySchedulerClient();
      final execs = await client.getExecutions('job-001');
      expect(execs, isEmpty);
    });

    test('スケジュールのバリアントが正しく生成されること', () {
      final cron = Schedule.cron('0 * * * *');
      expect(cron, isA<CronSchedule>());

      final oneShot = Schedule.oneShot(DateTime.now());
      expect(oneShot, isA<OneShotSchedule>());

      final interval = Schedule.interval(const Duration(minutes: 10));
      expect(interval, isA<IntervalSchedule>());
    });

    test('存在しないジョブのキャンセル時に例外がスローされること', () async {
      final client = InMemorySchedulerClient();
      expect(() => client.cancelJob('none'), throwsA(isA<SchedulerError>()));
    });

    test('ジョブ完了イベントが正しく生成されること', () {
      final event = JobCompletedEvent(
        jobId: 'job-1',
        executionId: 'exec-1',
        completedAt: DateTime.now(),
        result: 'success',
      );
      expect(event.jobId, equals('job-1'));
      expect(event.result, equals('success'));
    });
  });
}
